Adds a FeatureModelBuilder and iterates depth first

s9latimm opened the pull request:
- [x] FeatureModelBuilder
- [x] depth first iterator
- [x] comments
- [x] update bindings
- [x] clang-tidy

s9latimm committed:
WIP: Add builder

M	bindings/python/vara-feature/pybind_Location.cpp
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelParser.h
D	include/vara/Feature/Relationship.h
M	lib/Feature/Feature.cpp
M	lib/Feature/FeatureModelParser.cpp
M	tools/fm-viewer/FeatureModelViewer.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/CMakeLists.txt
A	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
WIP

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/CMakeLists.txt
M	lib/Feature/Feature.cpp
A	lib/Feature/FeatureModel.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
No dump

M	tools/fm-viewer/FeatureModelViewer.cpp

s9latimm committed:
WIP

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/FeatureSourceRange.h
M	lib/Feature/Feature.cpp
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/FeatureSourceRange.cpp

s9latimm committed:
Depth first iterator

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
A	include/vara/Feature/FeatureSort.h
M	lib/Feature/Feature.cpp
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/CMakeLists.txt
M	unittests/Feature/Feature.cpp
M	unittests/Feature/FeatureModel.cpp
A	unittests/Feature/FeatureSort.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
Revert to DenseSet

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -1,13 +1,15 @@
 #include "llvm/ADT/iterator_range.h"
 #include "llvm/Support/raw_ostream.h"
 
+#include <set>
```
seems like this is not used

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -20,28 +22,37 @@ namespace vara::feature {
 
 protected:
-  Feature(string Name, bool Opt, std::optional<Location> Loc)
-      : Name(std::move(Name)), Opt(Opt), Loc(std::move(Loc)) {}
+  Feature(string Name, bool Opt, std::optional<FeatureSourceRange> Loc,
+          Feature *Parent)
+      : Name(std::move(Name)), Opt(Opt), Loc(std::move(Loc)), Parent(Parent) {}
```
Protected section should come after public section

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -20,28 +22,37 @@ namespace vara::feature {
   bool Opt;
-  llvm::SmallVector<Feature *, 1> Parents;
-  FeatureListType Children;
-  FeatureListType Excludes;
-  FeatureListType Implications;
-  FeatureListType Alternatives;
-  RelationshipListTy Relationships;
+  Feature *Parent;
+  FeatureSetType Children;
+  FeatureSetType Excludes;
+  FeatureSetType Implications;
+  FeatureSetType Alternatives;
+
+  void addChild(Feature *F) { Children.insert(F); }
+
+  void setParent(Feature *F) { Parent = F; }
+
+  void addExclude(Feature *F) { Excludes.insert(F); }
+
+  void addAlternative(Feature *F) { Alternatives.insert(F); }
+
+  void addImplication(Feature *F) { Implications.insert(F); }
```
Private sections should come after public/protected sections

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -52,7 +63,11 @@ class Feature {
   [[nodiscard]] bool isOptional() const { return Opt; }
 
-  [[nodiscard]] bool isRoot() const { return Parents.empty(); }
+  [[nodiscard]] virtual bool isBinary() const = 0;
+
+  [[nodiscard]] virtual bool isNumeric() const = 0;
```
Why to we implement specific check functions when we use inheritance? If we need this information we should impl llvm style type info with kind enums. Then we can write `llvm::isa<BinaryFeature *>()`

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -203,30 +169,41 @@ class Feature {
-    return llvm::StringRef(this->getName()).lower() <
-           llvm::StringRef(F.getName()).lower();
+  struct FeatureDepthFirstComparator {
+    bool operator()(vara::feature::Feature *, vara::feature::Feature *) const;
```
Is a ptr here required? I would assume `const &` should be enough here

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -17,7 +19,8 @@ namespace vara::feature {
+  using ConstraintTy = llvm::SmallVector<std::pair<std::string, bool>, 3>;
+  using ConstraintsTy = std::vector<ConstraintTy>;
 
 private:
```
same public/protected/private order problem

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -68,29 +69,135 @@ class FeatureModel {
+  using feature_model_iterator = FeatureModelIter;
+  using const_feature_model_iterator = FeatureModelIter;
+
+  feature_model_iterator begin() { return FeatureModelIter(Features.begin()); }
```
Why is the FeatureModelIter private but exposed throught the public interface?

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -68,29 +69,135 @@ class FeatureModel {
 
+  [[nodiscard]] const_feature_model_iterator end() const {
+    return FeatureModelIter(Features.end());
+  }
```
Please add also add llvm style range function `features`

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -68,29 +69,135 @@ class FeatureModel {
+    using EdgeMapType =
+        typename llvm::StringMap<llvm::SmallVector<std::string, 3>>;
+
+    std::string VmName;
```
private/public order

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -68,29 +69,135 @@ class FeatureModel {
+      Children[P].push_back(C);
+      Parents[C] = P;
+      return this;
+    }
```
Does it make sense to also add overloads for when we have a `Feature &` and what to add a child etc.?

s9latimm replied:
This method / builder is supposed to be only used while parsing / iteratively building a FM, so it makes not that much sense to build those connections for an already existing `Feature` by hand or even hold such an object in the first place. I added `addFeature(Feature&)` instead which basically behaves like inserting all related connections one after another, which can be used to duplicate existing subtrees between different FMs.

vulder commented on include/vara/Feature/FeatureModelParser.h:
```diff
@@ -99,28 +99,15 @@ class FeatureModelXmlParser : public FeatureModelParser {
 
 private:
```
This private section should also be at the bottom

vulder commented on include/vara/Feature/FeatureSort.h:
```diff
@@ -0,0 +1,40 @@
+
+namespace vara::feature {
+
+class FeatureSort {
```
Sounds like an algorithm. I would prefer name like `OrderedFeatureVector`

vulder replied:
Currently, there are insertion methods missing. And I also do not see how this is used in the FM. I'll let you finish the impl before I'll continue to look at stuff :)

vulder commented on include/vara/Feature/FeatureSort.h:
```diff
@@ -0,0 +1,40 @@
+  using const_feature_heap_iterator =
+      typename std::vector<Feature *>::const_iterator;
+
+private:
```
private to the bottom

vulder commented on include/vara/Feature/FeatureSort.h:
```diff
@@ -0,0 +1,40 @@
+    for (auto F = FM->begin(); F != FM->end(); ++F) {
+      Features.push_back(*F);
+    }
+    std::sort(Features.begin(), Features.end(),
+              Feature::FeatureDepthFirstComparator());
+  }
```
Currently, this is not a heap. Didn't you want to make it a heap?

s9latimm replied:
Building a heap I would need to call `sort_heap` to receive a sorted collection while losing the heap property. As a custom iterator does not seem reasonable at this point, I used a  comparator with insertion at `std::upper_bound` instead.

vulder reviewed (changes requested):
sry, for the public/protected/private stuff I missed this last time

s9latimm committed:
Requested changes

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/FeatureSourceRange.h
M	lib/Feature/FeatureModel.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/CMakeLists.txt
M	unittests/Feature/Feature.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
Add range function

M	include/vara/Feature/FeatureModel.h

s9latimm committed:
Use ordered iterator

M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/OrderedFeatureVector.h
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/OrderedFeatureVector.cpp

s9latimm committed:
Improve comparator

M	include/vara/Feature/Feature.h
M	include/vara/Feature/OrderedFeatureVector.h
M	lib/Feature/Feature.cpp
M	unittests/Feature/Feature.cpp
M	unittests/Feature/OrderedFeatureVector.cpp

s9latimm committed:
Fix tidy

M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureSourceRange.h

s9latimm committed:
Merge branch 'vara-dev' into f-FMBuilderBindings

M	.gitignore
A	MANIFEST.in
M	bindings/python/tests/feature/test_feature.py
M	bindings/python/tests/feature/test_location.py
M	bindings/python/tests/feature_model/test_feature_model.py
M	bindings/python/vara-feature/CMakeLists.txt
M	bindings/python/vara-feature/pybind_Feature.cpp
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/FeatureSourceRange.h
M	lib/Feature/CMakeLists.txt
M	setup.py
M	test/vm.dtd
M	unittests/Feature/Feature.cpp
M	unittests/Feature/FeatureSourceRange.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
Add feature

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModel.cpp

s9latimm committed:
Add comments and fix minor issues

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/FeatureSourceRange.h
M	include/vara/Feature/OrderedFeatureVector.h
M	lib/Feature/Feature.cpp
M	lib/Feature/FeatureModel.cpp
M	unittests/Feature/Feature.cpp

s9latimm committed:
No move

M	include/vara/Feature/FeatureSourceRange.h

s9latimm committed:
Use const ref

M	include/vara/Feature/FeatureSourceRange.h

s9latimm committed:
NOLINT

M	include/vara/Feature/FeatureSourceRange.h

s9latimm committed:
Use const ref

M	include/vara/Feature/FeatureSourceRange.h

s9latimm committed:
NOLINT

M	include/vara/Feature/FeatureSourceRange.h

vulder commented on bindings/python/tests/feature/test_feature.py:
```diff
@@ -28,74 +29,18 @@ def test_is_root(self):
-        test_feature.add_alternative(root_feature)
-
-        self.assertTrue(test_feature.is_alternative(root_feature))
+    # def test_iter_children(self):
```
why is this case outcommented?

vulder commented on bindings/python/vara-feature/pybind_FeatureSourceRange.cpp:
```diff
@@ -0,0 +1,70 @@
+namespace vf = vara::feature;
+namespace py = pybind11;
+
+void init_feature_module_location(py::module &M) {
```
`init_feature_location_module`

vulder commented on bindings/python/vara-feature/pybind_FeatureSourceRange.cpp:
```diff
@@ -0,0 +1,70 @@
+           R"pbdoc(Get the start `LineColumnOffset` of this `Location`.)pbdoc")
+      .def("get_end", &vf::FeatureSourceRange::getEnd,
+           py::return_value_policy::reference,
+           R"pbdoc(Get the end `LineColumnOffset` of this `Location`.)pbdoc")
```
If I remember correctly pascal migrated accessors to python properties, we should use this here too.

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -21,44 +22,25 @@ namespace vara::feature {
-protected:
-  Feature(string Name, bool Opt, std::optional<Location> Loc)
-      : Name(std::move(Name)), Opt(Opt), Loc(std::move(Loc)) {}
+  enum FeatureType { BINARY, NUMERIC };
```
```suggestion
  enum FeatureKind { FK_Binary, FK_Numeric };
```

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -21,44 +22,25 @@ namespace vara::feature {
-  Feature &operator=(const Feature &) = delete;
   virtual ~Feature() = default;
 
+  [[nodiscard]] FeatureType getType() const { return T; }
```
```suggestion
  [[nodiscard]] FeatureKind getKind() const { return T; }
```

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -203,34 +136,90 @@ class Feature {
-  [[nodiscard]] Location* getLocation() {
+  //===--------------------------------------------------------------------===//
+  // Operators
+  Feature &operator=(const Feature &) = delete;
```
Copy assignment operator should be moved to the ctor if possible.

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -203,34 +136,90 @@ class Feature {
+  void dump() const { llvm::outs() << toString() << "\n"; }
+
+protected:
+  Feature(FeatureType T, string Name, bool Opt,
```
```suggestion
  Feature(FeatureKind Kind, string Name, bool Opt,
```

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -203,34 +136,90 @@ class Feature {
+
+  void addAlternative(Feature *F) { Alternatives.insert(F); }
+
+  void addImplication(Feature *F) { Implications.insert(F); }
```
```suggestion
  void addChild(Feature *F) { Children.insert(F); }

  void setParent(Feature *F) { Parent = F; }

  void addExclude(Feature *F) { Excludes.insert(F); }

  void addAlternative(Feature *F) { Alternatives.insert(F); }

  void addImplication(Feature *F) { Implications.insert(F); }
  
  FeatureType T;
  string Name;
  bool Opt;
  std::optional<FeatureSourceRange> Loc;
  Feature *Parent;
  FeatureSetType Children;
  FeatureSetType Excludes;
  FeatureSetType Implications;
  FeatureSetType Alternatives;
```
Nit: I would prefer to have all members at the end of the class.

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -203,34 +136,90 @@ class Feature {
+                FeatureSetType Alternatives = {})
+      : Feature(BINARY, std::move(Name), Opt, std::move(Loc), Parent,
+                std::move(Children), std::move(Excludes),
+                std::move(Implications), std::move(Alternatives)) {}
```
Does it make sense to make the ctor protected and only offer a factory function on the `FeatureModel`? Otherwise, we cannot guarantee that the values set in the ctor are consistent with the rest of the model.

s9latimm replied:
FM consistency is assured through that Features will only be added with `FeatureModel::addFeature(std::unique_ptr<Feature>) : bool` checking whether a Feature can be inserted without violating any rules (currently only rudimentary). On the other hand it may be handy to create (unrelated) features beforehand to store information (not only for testing), building the actual FM afterwards (as the builder provides corresponding methods).

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -1,37 +1,40 @@
+      : Name(std::move(Name)), RootPath(std::move(RootPath)), Root(Root),
+        Features(std::move(Features)) {
+    // Insert all values into ordered data structure.
+    for (const auto &KV : this->Features) {
+      OrderedFeatures.insert(KV.getValue().get());
+    }
```
Update to new insert method. (see below)

vulder commented on include/vara/Feature/FeatureSourceRange.h:
```diff
@@ -0,0 +1,102 @@
+  public:
+    FeatureSourceLocation(int Line, int Column) : Line(Line), Column(Column) {}
+
+    void setLineNumber(int L) { this->Line = L; }
```
```suggestion
    void setLineNumber(int LineNumber) { this->Line = LineNumber; }
```

vulder commented on include/vara/Feature/FeatureSourceRange.h:
```diff
@@ -0,0 +1,102 @@
+    void setLineNumber(int L) { this->Line = L; }
+    [[nodiscard]] int getLineNumber() const { return this->Line; }
+
+    void setColumnOffset(int C) { this->Column = C; }
```
```suggestion
    void setColumnOffset(int ColumnOffset) { this->Column = ColumnOffset; }
```

vulder commented on include/vara/Feature/OrderedFeatureVector.h:
```diff
@@ -0,0 +1,46 @@
+                         [](vara::feature::Feature *A,
+                            vara::feature::Feature *B) { return *A < *B; }),
+        F);
+  }
```
It would be helpful to provide different overloads for insert, especially, the one that takes and iterator range :wink: .

vulder commented on lib/Feature/Feature.cpp:
```diff
@@ -1,57 +1,34 @@
 #include "vara/Feature/Feature.h"
 
+#include <experimental/iterator>
```
```suggestion
#include <iterator>
```
Why would we need experimental stuff here?

vulder commented on unittests/Feature/Feature.cpp:
```diff
@@ -1,59 +1,79 @@
+  EXPECT_LT(*FM->getFeature("root"), *FM->getFeature("b"));
+  EXPECT_LT(*FM->getFeature("a"), *FM->getFeature("aa"));
+  EXPECT_LT(*FM->getFeature("a"), *FM->getFeature("b"));
+  EXPECT_LT(*FM->getFeature("aa"), *FM->getFeature("b"));
```
The less than impl tries to implement a quite complicated relationship, could you please split this case up into different test cases that cover the different scenarios.

vulder commented on lib/Feature/FeatureModel.cpp:
```diff
@@ -0,0 +1,148 @@
+    }
+  }
+
+  for (const auto &C : Constraints) {
```
If we iterate over lists here we should use a better name than `C`. What is the actual type of `C`? maybe it could help to spell it out instead of auto.

vulder commented on lib/Feature/FeatureModel.cpp:
```diff
@@ -0,0 +1,148 @@
+          }
+        }
+      }
+    }
```
What happens in the else case? If `C.size()` is `0` or `1`?

vulder commented on lib/Feature/Feature.cpp:
```diff
@@ -78,4 +52,45 @@ std::string NumericFeature::toString() const {
 
 std::string BinaryFeature::toString() const { return Feature::toString(); }
 
+bool Feature::operator<(const vara::feature::Feature &F) const {
```
For a `operator<` this is quite complex, so we should test this well, do the test cover all edge cases?

s9latimm replied:
I tested all pairwise relations on binary trees with height 2, basic cases and also unrelated features.

vulder reviewed (changes requested):
Found a few minor things

s9latimm committed:
Update lib/Feature/Feature.cpp

Co-authored-by: Florian Sattler <vuld3r@gmail.com>

s9latimm committed:
Refactoring

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelParser.h
M	lib/Feature/Feature.cpp
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/FeatureModel.cpp

s9latimm committed:
Improve and test comparison

M	lib/Feature/Feature.cpp
M	unittests/Feature/Feature.cpp

s9latimm committed:
More tests

M	unittests/Feature/Feature.cpp

s9latimm committed:
Add insert methods

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureSourceRange.h
M	include/vara/Feature/OrderedFeatureVector.h
M	lib/Feature/CMakeLists.txt
M	unittests/Feature/OrderedFeatureVector.cpp

s9latimm committed:
Add insert methods

A	lib/Feature/OrderedFeatureVector.cpp

s9latimm committed:
Add builder tests

M	bindings/python/vara-feature/pybind_Feature.cpp
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/FeatureModel.cpp
M	unittests/Feature/CMakeLists.txt
M	unittests/Feature/FeatureModel.cpp
A	unittests/Feature/FeatureModelBuilder.cpp

s9latimm committed:
Add ternary test

M	lib/Feature/FeatureModel.cpp
M	unittests/Feature/Feature.cpp

s9latimm committed:
Update bindings

M	bindings/python/tests/feature/test_feature.py
M	bindings/python/tests/feature/test_location.py
M	bindings/python/vara-feature/pybind_Feature.cpp
M	bindings/python/vara-feature/pybind_FeatureSourceRange.cpp
M	bindings/python/vara-feature/pybind_Init.cpp
M	include/vara/Feature/Feature.h
M	lib/Feature/FeatureModel.cpp

vulder commented on bindings/python/vara-feature/pybind_Feature.cpp:
```diff
@@ -55,42 +47,37 @@ void init_feature_module_feature(py::module &M) {
-          "add_implication", &vf::Feature::addImplication,
-          R"pbdoc(Implicate another Feature, i.e., selecting this feature implies the otherone.)pbdoc")
-      .def("implicates", &vf::Feature::implies,
+      .def("is_implied", &vf::Feature::isImplied,
```
```suggestion
      .def("is_implied_by", &vf::Feature::isImplied,
```
This slight name change should make it a bit clearer

vulder commented on unittests/Feature/OrderedFeatureVector.cpp:
```diff
@@ -0,0 +1,41 @@
+
+namespace vara::feature {
+
+TEST(OrderedFeatureVector, insert) {
+  auto FM = FeatureModelBuilder().buildSimpleFeatureModel(
+      {{"root", "a"}, {"root", "b"}});
+  OrderedFeatureVector OFV;
+
+  OFV.insert(FM->getFeature("a"));
+  OFV.insert(FM->getFeature("root"));
+  OFV.insert(FM->getFeature("b"));
+
+  EXPECT_EQ(OFV.size(), FM->size());
+  EXPECT_EQ(*OFV.begin(), FM->getFeature("root"));
+}
+
+TEST(OrderedFeatureVector, insertFM) {
+  auto FM = FeatureModelBuilder().buildSimpleFeatureModel(
+      {{"root", "a"}, {"root", "b"}});
+  OrderedFeatureVector OFV;
+
+  OFV.insert(FM->features());
+
+  EXPECT_EQ(OFV.size(), FM->size());
+  EXPECT_EQ(*OFV.begin(), FM->getFeature("root"));
+}
+
+TEST(OrderedFeatureVector, insertVariadic) {
+  auto FM = FeatureModelBuilder().buildSimpleFeatureModel(
+      {{"root", "a"}, {"root", "b"}});
+  OrderedFeatureVector OFV;
+
+  OFV.insert(FM->getFeature("a"), FM->getFeature("root"), FM->getFeature("b"));
+
+  EXPECT_EQ(OFV.size(), FM->size());
+  EXPECT_EQ(*OFV.begin(), FM->getFeature("root"));
```
All tests only check if the first one is root but do not verify if the others are in correct order.

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -1,96 +1,206 @@
 
-  FeatureModelIter begin() { return FeatureModelIter(Features.begin()); }
+  /// Try to add \a BinaryFeature.
+  bool addFeature(const std::string &Key, bool Opt = false,
+                  std::optional<FeatureSourceRange> Loc = std::nullopt) {
+    return Features
+        .try_emplace(Key,
+                     std::make_unique<BinaryFeature>(Key, Opt, std::move(Loc)))
+        .second;
+  }
 
-  FeatureModelIter end() { return FeatureModelIter(Features.end()); }
+  /// Try to add \a NumericFeature.
+  bool addFeature(const std::string &Key,
+                  const NumericFeature::ValuesVariantType &Values,
+                  bool Opt = false,
+                  std::optional<FeatureSourceRange> Loc = std::nullopt) {
+    return Features
+        .try_emplace(Key, std::make_unique<NumericFeature>(Key, Values, Opt,
+                                                           std::move(Loc)))
+        .second;
+  }
```
We should merge these functions into a factory method and rename them to `make_feature` or `create_feature`, to not be confused with the `addFeature(Feature &)` method, which has a different behavior (Add a feature by copying it vs creating a new feature). Furthermore, `Key` is not a really good name here, it is the key for the map but from the API users side it is the feature name, and we should not confuse the user by naming things based on our implementation details.

Regarding the factory method:
```
    /// Try to create and add a new \a Feature to the \a FeatureModel.
  ///
  /// \param[in] FeatureName name of the \a Feature
  /// \param[in] FurtherArgs further arguments that should be passed to the
  ///                        \a Feature constructor
  ///
  /// \returns true, if the feature could be inserted into the \a FeatureModel
  template <
      typename FeatureTy, typename... Args,
      typename std::enable_if_t<std::is_base_of_v<Feature, FeatureTy>, int>>
  bool makeFeature(const std::string &FeatureName, Args... FurtherArgs) {
    return Features
        .try_emplace(FeatureName, std::make_unique<FeatureTy>(
                                      FeatureName, std::forward(FurtherArgs)...))
        .second;
  }
```
This allows users to create any kind of feature as long as it is derived from `Feature` and the passed args match one of the specific Features ctors.

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -1,96 +1,206 @@
-    }
+  bool addFeature(Feature &F);
+
+  FeatureModelBuilder *addParent(const std::string &Key,
+                                 const std::string &ParentKey) {
+    Children[ParentKey].insert(Key);
+    Parents[Key] = ParentKey;
+    return this;
+  }
+
+  FeatureModelBuilder *addExclude(const std::string &Key,
+                                  const std::string &ExcludeKey) {
+    Excludes[Key].insert(ExcludeKey);
+    return this;
+  }
```
What happens in these functions if the `Key` is not a Feature name that was inserted before?

PS: `Key` is also a bad name for this param, we should rename this.

s9latimm replied:
Doing so is totally fine, as the consistency (existence) is asserted when building the final FM. As the parser can not assure that features are only referenced if they are already processed (like mutual exclusion), this can not be asserted in this method either.

vulder commented on include/vara/Feature/OrderedFeatureVector.h:
```diff
@@ -0,0 +1,51 @@
+  OrderedFeatureVector() = default;
+
+  /// Insert feature while preserving ordering.
+  void insert(Feature *F);
```
```suggestion
  void insert(Feature *F);
  
```

vulder commented on include/vara/Feature/OrderedFeatureVector.h:
```diff
@@ -0,0 +1,51 @@
+    insert(F);
+    insert(FF...);
+  }
+  template <typename T> void insert(llvm::iterator_range<T> FF) {
```
```suggestion

  template <typename T> void insert(llvm::iterator_range<T> FF) {
```

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -21,44 +23,26 @@ namespace vara::feature {
-protected:
-  Feature(string Name, bool Opt, std::optional<Location> Loc)
-      : Name(std::move(Name)), Opt(Opt), Loc(std::move(Loc)) {}
+  enum FeatureKind { FK_BINARY, FK_NUMERIC };
```
Maybe it would be better to use a enum class here so integer constant can not be confused with the feature kinds.

vulder reviewed (changes requested):
Test coverage looks much better now :smiley:

s9latimm committed:
Apply suggestions from code review

Co-authored-by: Florian Sattler <vuld3r@gmail.com>

s9latimm committed:
Requested changes

M	bindings/python/tests/feature/test_feature.py
M	bindings/python/vara-feature/pybind_Feature.cpp
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/FeatureModel.cpp
M	unittests/Feature/Feature.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/OrderedFeatureVector.cpp

s9latimm committed:
Add factory method call

M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/OrderedFeatureVector.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/NumericFeature.cpp

vulder reviewed (approved):
lgtm
